test/membrane/error.spec.js

```diff
@@ -148,14 +139,164 @@ it('non-error objects thrown in blue constructors do not leak', () => {
         try {
             new Foo()
         } catch (e) {
-            // blue error own properties on thrown object do not leak
             expect(e.message).toBe(undefined)
-            expect(e.foo).toBe(undefined)
+            expect(e.foo).toBe('bar')
+        }
+    `)
+})
+
+it('[red] blue extended error objects', () => {
+    class CustomError extends Error {
+        constructor(message) {
+            super(message)
+            this.bar = 'baz'
+        }
+
+        get foo() {
+            return 'bar'
+        }
+    }
+
+    class Foo {
+        constructor() {
+            throw new CustomError('foo');
+        }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { Foo, expect })
+    evalScript(`
+        try {
+            new Foo()
+        } catch (e) {
+            expect(e.foo).toBe('bar')
+            expect(e.bar).toBe('baz')
+            expect(e.message).toBe('foo')
+        }
+    `)
+})
+
+it('[red] .catch on blue promise', (done) => {
+    const promise = new Promise(() => {
+        throw {foo: 'bar'}
+    })
+
+    const evalScript = createSecureEnvironment(undefined, { promise, expect, done })
+    evalScript(`
+        promise.catch(e => {
+            expect(e.foo).toBe('bar')
+            expect(e.message).toBe(undefined)
+            done()
+        })
+    `)
+})
+
+it('[red] non-error object with null proto', () => {
+    const evalScript = createSecureEnvironment(undefined, { expect })
+    evalScript(`
+        const errorObj = Object.create(null, {foo: {value: 'bar'}})
+        try {
+            throw errorObj
+        } catch(e) {
+            expect(errorObj).toBe(errorObj)
+            expect(Reflect.getPrototypeOf(errorObj)).toBe(null)
+            expect(errorObj.message).toBe(undefined)
+            expect(errorObj.foo).toBe('bar')
+        }
+    `)
+})
+
+it('[red] non-error object with null proto from blue', () => {
+    function foo() {
+        throw Object.create(null, {foo: {value: 'bar'}})
+    }
+    const evalScript = createSecureEnvironment(undefined, { foo, expect })
+    evalScript(`
+        try {
+            foo()
+        } catch(e) {
+            expect(Reflect.getPrototypeOf(e)).toBe(null)
+            expect(e.message).toBe(undefined)
+            expect(e.foo).toBe('bar')
         }
     `)
 })
 
```

Comment:
The below test is an interesting scenario and it's something that definitely needs a conversation. Right now the sandbox is able to intercept errors from blue. Should the sandbox protect this scenario by default? Or should this be addressed via a distortion on window.addEventListener ?